### PR TITLE
listener: fix multiple issues around creation failures

### PR DESCRIPTION
**Pull request by mattklein123**
Previously we were not correctly handling listener creation failures
when the failure occurred on the worker thread (as in the case of
using reuse port).

Fixes https://github.com/envoyproxy/envoy/issues/11340
Fixes https://github.com/envoyproxy/envoy/issues/10070

Risk Level: Low
Testing: New/existing tests
Docs Changes: N/A
Release Notes: Added

### CODE REVIEW

**Comment by mattklein123** [comment]
@danzh2010 @ggreenway PTAL

**Comment by ggreenway** [comment]
Would the error handling be easier if we create and bind the listen socket on the main thread, then pass that to the worker to do `evconnlistener_new`?

File: test/server/connection_handler_test.cc [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #include "envoy/config/core/v3/base.pb.h"
 #include "envoy/config/listener/v3/udp_listener_config.pb.h"
+#include "envoy/network/exception.h"
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply test/server/connection_handler_test.cc:3]: These are needed as the tests use the defined exceptions?

**Comment by mattklein123** [comment]
> Would the error handling be easier if we create and bind the listen socket on the main thread, then pass that to the worker to do evconnlistener_new?

Possibly, but this would be a very large change in the face of reuse_port, as we would need to pre-allocate sockets for all workers and then hand them out. But I agree that it would simplify things and possibly even allow for direct NACK of listener failures in certain cases. I can add a TODO/comment around this?

**Comment by ggreenway** [comment]
> > Would the error handling be easier if we create and bind the listen socket on the main thread, then pass that to the worker to do evconnlistener_new?
> 
> Possibly, but this would be a very large change in the face of reuse_port, as we would need to pre-allocate sockets for all workers and then hand them out. But I agree that it would simplify things and possibly even allow for direct NACK of listener failures in certain cases. I can add a TODO/comment around this?

Yes, a TODO is fine. It's out of scope for this change. But the NACK behavior is what makes it appealing to me.

**Comment by mattklein123** [comment]
> Yes, a TODO is fine. It's out of scope for this change. But the NACK behavior is what makes it appealing to me.

Yes agreed, will do.

**Comment by mattklein123** [comment]
@ggreenway sorry need another +1 due to format fix.

**Pull request merged**